src/storage/data_object_storage_registry.rs

```diff
@@ -0,0 +1,150 @@
+        pub DataObjectStorageRelationshipMap get(data_object_storage_relationship): map T::DataObjectStorageRelationshipId => Option<DataObjectStorageRelationship<T>>;
+
+        // Keep a list of storage relationships per CID
+        pub DataObjectStorageRelationships get(data_object_storage_relationships): map T::ContentId => Vec<T::DataObjectStorageRelationshipId>;
+    }
+}
+
+decl_event! {
+    pub enum Event<T> where
+        <T as DDTrait>::ContentId,
+        <T as Trait>::DataObjectStorageRelationshipId,
+        <T as system::Trait>::AccountId
+    {
+        DataObjectStorageRelationshipAdded(DataObjectStorageRelationshipId, ContentId, AccountId),
+        DataObjectStorageRelationshipReadyUpdated(DataObjectStorageRelationshipId, bool),
+    }
+}
+
+impl<T: Trait> ContentHasStorage<T> for Module<T> {
+    fn has_storage_provider(which: &T::ContentId) -> bool {
+        let dosr_list = Self::data_object_storage_relationships(which);
+        return dosr_list.iter().any(|&dosr_id| Self::data_object_storage_relationship(dosr_id).unwrap().ready);
+    }
+
+    fn is_ready_at_storage_provider(which: &T::ContentId, provider: &T::AccountId) -> bool {
+        let dosr_list = Self::data_object_storage_relationships(which);
+        return dosr_list.iter().any(|&dosr_id| {
+            let dosr = Self::data_object_storage_relationship(dosr_id).unwrap();
+            dosr.storage_provider == *provider && dosr.ready
+        });
+    }
+}
+
+
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        fn deposit_event<T>() = default;
+
+        pub fn add_data_object_storage_relationship(origin, cid: T::ContentId) {
+            // Origin has to be a storage provider
+            let who = ensure_signed(origin).clone().unwrap();
+            // TODO check for being staked as a storage provider
+            // if !T::IsActiveMember::is_active_member(&who) {
+            //     return Err(MSG_CREATOR_MUST_BE_MEMBER);
+            // }
+
+            // Content ID must exist
+            if !T::ContentIdExists::has_content(&cid) {
+                return Err(MSG_CID_NOT_FOUND);
+            }
+
+            // Create new ID, data.
+            let new_id = Self::next_data_object_storage_relationship_id();
+            let dosr: DataObjectStorageRelationship<T> = DataObjectStorageRelationship {
+                content_id: cid.clone(),
+                storage_provider: who.clone(),
+                ready: false,
+            };
+
+            <DataObjectStorageRelationshipMap<T>>::insert(new_id, dosr);
+            <NextDataObjectStorageRelationshipId<T>>::mutate(|n| { *n += T::DataObjectStorageRelationshipId::sa(1); });
+
+            // Also add the DOSR to the list of DOSRs for the CID. Uniqueness is guaranteed
+            // by the map, so we can just append the new_id to the list.
+            let mut dosr_list = Self::data_object_storage_relationships(cid.clone());
+            dosr_list.push(new_id);
```

Comment:
dosr_list is mutated but not saved back into storage

Reply:
Huh... good catch!

Reply:
Let the tests cover this? ;)